modules/diskq/diskq-global-metrics.c

```diff
@@ -169,6 +173,66 @@ _init_abandoned_disk_buffer_sc_keys(StatsClusterKey *queued_sc_key, StatsCluster
   stats_cluster_single_key_add_unit(disk_usage_sc_key, SCU_KIB);
 }
 
+static void
+_set_abandoned_disk_buffer_file_metrics(StatsClusterKey *queued_sc_key, StatsClusterKey *capacity_sc_key,
+                                        StatsClusterKey *disk_allocated_sc_key, StatsClusterKey *disk_usage_sc_key,
+                                        const gchar *dir, const gchar *abs_filename, gboolean reliable)
+{
+  DiskQueueOptions options;
+  disk_queue_options_set_default_options(&options);
+  disk_queue_options_set_dir(&options, dir);
+  options.reliable = reliable;
+  options.read_only = TRUE;
+
+  LogQueue *queue;
+  if (reliable)
+    queue = log_queue_disk_reliable_new(&options, abs_filename, NULL, STATS_LEVEL0, NULL, NULL);
+  else
+    queue = log_queue_disk_non_reliable_new(&options, abs_filename, NULL, STATS_LEVEL0, NULL, NULL);
+  LogQueueDisk *queue_disk = (LogQueueDisk *) queue;
+
+  if (!log_queue_disk_start(queue))
```

Comment:
Questions:
I'm wondering how expensive this call can be.

Is it okay to enable this by default? If so, is it OK to repeat it every 5 minutes?
A dir might have thousands of abandoned disk-buffer files (or not-abandoned, or just regular files).

This feature implies that 2 syslog-ng instances can not use the same disk-buffer directory, otherwise they will report each others' buffers as abandoned.
I think this is good behavior, but considering the above, I think it is unnecessary to do periodic abandoned buffer checks, because no one will change those values between an acquire/release cycle.

TLDR: I think we should remove the periodic abandoned buffer metrics update from the timer callback, and call it only when a new directory is added by `acquire`. This would keep this potentially expensive call at startup time.
What about buffers that were abandoned during runtime? I think it's not possible unless they are corrupted, but we should not try to open corrupted buffer files in the main syslog-ng process. Reloading might abandon buffers, but that will be detected in `acquire`.

Reply:
I agree. The only case where this module does not get notified about a new disk-buffer file is corruption. The code currently tries to open the corrupted file and if it succeeds it creates metrics for it.
Two things:
  1. We can call the `diskq_global_metrics_file_released()` fn in the corrupted restart fn, so we don't need the poll at all.
  2. As you mentioned, we might not want to create metrics for the corrupted files. But I would ask you what is the reasoning behind it? I did not see a problem with it.

Reply:
I'm afraid a corrupted disk-buffer file may crash the syslog-ng instance. Also, corrupt files may have corrupt headers, therefore invalid entries.

I know... Every crash possibility should be fixed.
But if we could avoid doing all of this at runtime (moving them to startup time only), I think it would be much safer and cheaper. A metric, after all, should not affect performance or stability.

I'd vote for 2), but if you insist, 1) is good too.

Reply:
I have added a check to skip the corrupted files.

Reply:
Thank you.